File: frontend/src/api/routing.ts
export type CoordinateInput = {
  lat: number
  lng: number
}

export type DirectedEdgeReference = {
  edge_id: string
  u: string
  v: string
  key: number
  road_name: string
  road_class: string
  osm_way_ids: string[]
  geometry_fingerprint: string
  lanes: number
  maxspeed_kph: number
}

export type SelectedLocation = CoordinateInput & {
  node_id: string
  distance_m: number
  label: string
  edge: DirectedEdgeReference
}

export type RoadSelectionResponse = {
  graph_version: string
  location: SelectedLocation
}

export type ClosureDirectionCandidate = {
  edge: DirectedEdgeReference
  direction_label: string
  geometry: {
    type: 'LineString'
    coordinates: [number, number][]
  }
}

export type ClosureRoadSelectionResponse = {
  graph_version: string
  road_name: string
  distance_m: number
  selected_edge_id: string
  directions: ClosureDirectionCandidate[]
}

export type RouteSummary = {
  route_id: string
  travel_time_seconds: number
  distance_m: number
  geometry: {
    type: 'LineString'
    coordinates: [number, number][]
  }
}

export type RouteCompareResponse = {
  graph_version: string
  evidence_level: 'free_flow' | 'modeled_uncalibrated'
  baseline: RouteSummary
  scenario: RouteSummary | null
  scenario_status: 'not_requested' | 'available' | 'no_route' | 'inactive'
  evaluated_departure_time: string
  applied_restriction_edge_ids: string[]
  inactive_restriction_edge_ids: string[]
  assumptions: string[]
}

export type AlternativeRoute = {
  ranking: 'fastest' | 'reliable' | 'balanced'
  route: RouteSummary
  overlap_with_fastest_percent: number
  residential_distance_percent: number
  description: string
}

export type RouteAlternativesResponse = {
  graph_version: string
  evidence_level: 'modeled_uncalibrated'
  routes: AlternativeRoute[]
  evaluated_departure_time: string
  assumptions: string[]
}

export type GoogleMapsUrlResponse = {
  url: string
  waypoint_count: number
  warning: string
}

export type RoadRestriction =
  | { type: 'full' }
  | { type: 'lane'; remaining_lanes: number }
  | { type: 'speed'; speed_limit_kph: number }

export type RestrictionSchedule =
  | { type: 'always' }
  | { type: 'scheduled'; starts_at: string; ends_at: string }

export type ClosureSectionDraft = {
  selection: ClosureRoadSelectionResponse
  restriction: RoadRestriction
  schedule: RestrictionSchedule
}

export type ClosureComparisonSection = {
  directions: ClosureDirectionCandidate[]
  restriction: RoadRestriction
  schedule: RestrictionSchedule
}

export type ClosureComparisonInput = {
  graphVersion: string
  departureTime: string
  sections: ClosureComparisonSection[]
}

export class ApiError extends Error {
  code: string

  constructor(code: string, message: string) {
    super(message)
    this.name = 'ApiError'
    this.code = code
  }
}

export async function selectRoad(
  coordinate: CoordinateInput,
  signal?: AbortSignal,
): Promise<RoadSelectionResponse> {
  return requestJson('/api/roads/select', coordinate, signal)
}

export async function selectClosureRoad(
  coordinate: CoordinateInput,
  signal?: AbortSignal,
): Promise<ClosureRoadSelectionResponse> {
  return requestJson('/api/roads/select-closure', coordinate, signal)
}

export async function compareRoute(
  origin: SelectedLocation,
  destination: SelectedLocation,
  closure?: ClosureComparisonInput,
  signal?: AbortSignal,
): Promise<RouteCompareResponse> {
  return requestJson(
    '/api/routes/compare',
    buildRouteRequestBody(origin, destination, closure),
    signal,
  )
}

export async function findAlternativeRoutes(
  origin: SelectedLocation,
  destination: SelectedLocation,
  closure?: ClosureComparisonInput,
  signal?: AbortSignal,
): Promise<RouteAlternativesResponse> {
  return requestJson(
    '/api/routes/alternatives',
    buildRouteRequestBody(origin, destination, closure),
    signal,
  )
}

export async function createGoogleMapsUrl(
  origin: SelectedLocation,
  destination: SelectedLocation,
  route: RouteSummary,
  signal?: AbortSignal,
): Promise<GoogleMapsUrlResponse> {
  return requestJson(
    '/api/routes/google-maps-url',
    {
      origin: { lat: origin.lat, lng: origin.lng },
      destination: { lat: destination.lat, lng: destination.lng },
      route,
    },
    signal,
  )
}

export function buildRouteRequestBody(
  origin: SelectedLocation,
  destination: SelectedLocation,
  closure?: ClosureComparisonInput,
) {
  return {
    origin: {
      lat: origin.lat,
      lng: origin.lng,
      node_id: origin.node_id,
    },
    destination: {
      lat: destination.lat,
      lng: destination.lng,
      node_id: destination.node_id,
    },
    graph_version: closure?.graphVersion,
    departure_time: closure?.departureTime,
    closures: closure
      ? closure.sections.map(({ directions, restriction, schedule }) => ({
          ...restriction,
          ...(schedule.type === 'scheduled'
            ? {
                starts_at: new Date(schedule.starts_at).toISOString(),
                ends_at: new Date(schedule.ends_at).toISOString(),
              }
            : {}),
          edges: directions.map(({ edge }) => ({
            edge_id: edge.edge_id,
            u: edge.u,
            v: edge.v,
            key: edge.key,
          })),
        }))
      : [],
  }
}

async function requestJson<Response>(
  url: string,
  body: unknown,
  signal?: AbortSignal,
): Promise<Response> {
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal,
  })

  if (!response.ok) {
    const payload = (await response.json().catch(() => null)) as {
      detail?: { code?: string; message?: string }
    } | null
    throw new ApiError(
      payload?.detail?.code ?? 'request_failed',
      payload?.detail?.message ?? 'The local routing request failed.',
    )
  }

  return response.json() as Promise<Response>
}

File: backend/app/schemas/routing.py
"""Typed map selection, closure, and routing API schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class Coordinate(BaseModel):
    """A WGS84 coordinate supplied or returned through the API."""

    lat: float = Field(ge=-90, le=90)
    lng: float = Field(ge=-180, le=180)


class RoadSelectionRequest(Coordinate):
    """Map click to snap against the local routable network."""


class DirectedEdgeReference(BaseModel):
    """Stable identity for the closest directed road edge."""

    edge_id: str
    u: str
    v: str
    key: int
    road_name: str
    road_class: str
    osm_way_ids: list[str] = Field(default_factory=list)
    geometry_fingerprint: str = ""
    lanes: float = Field(gt=0)
    maxspeed_kph: float = Field(gt=0)


class SnappedLocation(Coordinate):
    """A map point resolved to a routable graph node and nearby road."""

    node_id: str
    distance_m: float = Field(ge=0)
    label: str
    edge: DirectedEdgeReference


class RoadSelectionResponse(BaseModel):
    """Confirmed local-network selection for the frontend workflow."""

    graph_version: str
    location: SnappedLocation


class ClosureDirectionCandidate(BaseModel):
    """One directed edge a user can explicitly close."""

    edge: DirectedEdgeReference
    direction_label: str
    geometry: "GeoJSONLineString"


class ClosureRoadSelectionResponse(BaseModel):
    """The nearest road and each independently closable direction."""

    graph_version: str
    road_name: str
    distance_m: float = Field(ge=0)
    selected_edge_id: str
    directions: list[ClosureDirectionCandidate] = Field(min_length=1)


class RouteLocation(Coordinate):
    """A trip endpoint, optionally retaining a previously snapped node."""

    node_id: str | None = None


class ClosureEdgeInput(BaseModel):
    """Graph-version-specific directed edge selected by the user."""

    edge_id: str
    u: str
    v: str
    key: int


class ClosureInput(BaseModel):
    """One restriction applied to explicitly selected directed edges."""

    type: Literal["full", "lane", "speed"] = "full"
    edges: list[ClosureEdgeInput] = Field(min_length=1)
    remaining_lanes: float | None = Field(default=None, gt=0)
    speed_limit_kph: float | None = Field(default=None, ge=5, le=130)
    starts_at: datetime | None = None
    ends_at: datetime | None = None

    @model_validator(mode="after")
    def validate_restriction_value(self) -> "ClosureInput":
        """Require only the value used by the selected restriction type."""

        if self.type == "lane" and self.remaining_lanes is None:
            raise ValueError("Lane restrictions require remaining_lanes.")
        if self.type == "speed" and self.speed_limit_kph is None:
            raise ValueError("Speed restrictions require speed_limit_kph.")
        if self.type != "lane" and self.remaining_lanes is not None:
            raise ValueError("remaining_lanes is only valid for lane restrictions.")
        if self.type != "speed" and self.speed_limit_kph is not None:
            raise ValueError("speed_limit_kph is only valid for speed restrictions.")
        if (self.starts_at is None) != (self.ends_at is None):
            raise ValueError("Scheduled restrictions require both starts_at and ends_at.")
        if self.starts_at is not None and self.ends_at is not None:
            if not _is_timezone_aware(self.starts_at) or not _is_timezone_aware(
                self.ends_at
            ):
                raise ValueError("Restriction schedules must include a timezone.")
            if self.ends_at <= self.starts_at:
                raise ValueError("Restriction end time must be after its start time.")
        return self


class RouteCompareRequest(BaseModel):
    """Trip request with optional directed road restrictions."""

    origin: RouteLocation
    destination: RouteLocation
    graph_version: str | None = None
    departure_time: datetime | None = None
    closures: list[ClosureInput] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_departure_time(self) -> "RouteCompareRequest":
        if self.departure_time is not None and not _is_timezone_aware(
            self.departure_time
        ):
            raise ValueError("departure_time must include a timezone.")
        return self


class GeoJSONLineString(BaseModel):
    """Small route geometry for MapLibre display."""

    type: Literal["LineString"] = "LineString"
    coordinates: list[tuple[float, float]] = Field(min_length=2)

    @model_validator(mode="after")
    def validate_coordinates(self) -> "GeoJSONLineString":
        for longitude, latitude in self.coordinates:
            if not -180 <= longitude <= 180 or not -90 <= latitude <= 90:
                raise ValueError("Route geometry contains an invalid coordinate.")
        return self


class RouteSummary(BaseModel):
    """A free-flow route with honest evidence and display metrics."""

    route_id: str
    travel_time_seconds: float = Field(ge=0)
    distance_m: float = Field(ge=0)
    geometry: GeoJSONLineString


class RouteCompareResponse(BaseModel):
    """Baseline and closure-aware results from the same immutable graph."""

    graph_version: str
    evidence_level: Literal["free_flow", "modeled_uncalibrated"]
    baseline: RouteSummary
    scenario: RouteSummary | None = None
    scenario_status: Literal["not_requested", "available", "no_route", "inactive"] = (
        "not_requested"
    )
    evaluated_departure_time: datetime
    applied_restriction_edge_ids: list[str] = Field(default_factory=list)
    inactive_restriction_edge_ids: list[str] = Field(default_factory=list)
    assumptions: list[str]


class AlternativeRoute(BaseModel):
    """One distinct corridor with transparent ranking inputs."""

    ranking: Literal["fastest", "reliable", "balanced"]
    route: RouteSummary
    overlap_with_fastest_percent: float = Field(ge=0, le=100)
    residential_distance_percent: float = Field(ge=0, le=100)
    description: str


class RouteAlternativesResponse(BaseModel):
    """Meaningfully distinct closure-aware route choices."""

    graph_version: str
    evidence_level: Literal["modeled_uncalibrated"] = "modeled_uncalibrated"
    routes: list[AlternativeRoute] = Field(min_length=1, max_length=3)
    evaluated_departure_time: datetime
    assumptions: list[str]


class GoogleMapsUrlRequest(BaseModel):
    """A backend-produced route and its trip endpoints for handoff."""

    origin: Coordinate
    destination: Coordinate
    route: RouteSummary


class GoogleMapsUrlResponse(BaseModel):
    """Standard unkeyed Google Maps directions URL and disclosure."""

    url: str
    waypoint_count: int = Field(ge=0, le=3)
    warning: str


def _is_timezone_aware(value: datetime) -> bool:
    return value.tzinfo is not None and value.utcoffset() is not None
